File: backend-fastapi/tools/contracts/permissions.py
"""Pure permission models for the tool system."""

from enum import Enum

from pydantic import BaseModel


class RiskLevel(str, Enum):
    """工具风险等级"""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ToolPermission(BaseModel):
    """工具权限配置"""

    tool_name: str
    risk_level: RiskLevel = RiskLevel.LOW
    requires_approval: bool = False
    description: str = ""
    allowed_roles: list = []
    allowed_callers: list = ["direct", "code_execution"]
    timeout_seconds: int = 60

File: backend-ts/src/contracts/permissions.ts
import { z } from "zod";

export const PermissionModeSchema = z.enum([
  "strict",
  "standard",
  "relaxed",
  "dangerously_skip_permissions",
]);

export const RiskLevelSchema = z.enum(["low", "medium", "high"]);

export const AutoAcceptPatternSchema = z.object({
  pattern_type: z.string().min(1),
  pattern_value: z.string().min(1),
  description: z.string().optional().default(""),
});

export const ToolPermissionSchema = z.object({
  tool_name: z.string().min(1),
  risk_level: RiskLevelSchema.optional().default("low"),
  description: z.string().optional().default(""),
  allowed_roles: z.array(z.string()).optional().default([]),
  allowed_callers: z.array(z.string()).optional().default(["direct", "code_execution"]),
  timeout_seconds: z.number().int().positive().optional().default(60),
});

export const PermissionPolicySchema = z.object({
  mode: PermissionModeSchema.optional().default("standard"),
  auto_accept_patterns: z.array(AutoAcceptPatternSchema).optional().default([]),
  audit_all_checks: z.boolean().optional().default(false),
  approval_timeout: z.number().int().positive().optional().default(300),
  skip_all_approvals: z.boolean().optional().default(false),
});

export type PermissionMode = z.infer<typeof PermissionModeSchema>;
export type RiskLevel = z.infer<typeof RiskLevelSchema>;
export type AutoAcceptPattern = z.infer<typeof AutoAcceptPatternSchema>;
export type ToolPermission = z.infer<typeof ToolPermissionSchema>;
export type PermissionPolicy = z.infer<typeof PermissionPolicySchema>;
